File: src/brs/db/store/AccountStore.java
package brs.db.store;

import brs.Account;
import brs.Asset;
import brs.db.SignumKey;
import brs.db.VersionedBatchEntityTable;
import brs.db.VersionedEntityTable;

import java.util.Collection;

/**
 * Interface for Database operations related to Accounts
 */
public interface AccountStore {

  VersionedBatchEntityTable<Account> getAccountTable();

  VersionedBatchEntityTable<Account.Balance> getAccountBalanceTable();

  long getAllAccountsBalance();

  VersionedEntityTable<Account.RewardRecipientAssignment> getRewardRecipientAssignmentTable();

  SignumKey.LongKeyFactory<Account.RewardRecipientAssignment> getRewardRecipientAssignmentKeyFactory();

  SignumKey.LinkKeyFactory<Account.AccountAsset> getAccountAssetKeyFactory();

  VersionedEntityTable<Account.AccountAsset> getAccountAssetTable();

  int getAssetAccountsCount(Asset asset, long minimumQuantity, boolean ignoreTreasury, boolean unconfirmed);

  long getAssetCirculatingSupply(Asset asset, boolean ignoreTreasury, boolean unconfirmed);

  SignumKey.LongKeyFactory<Account> getAccountKeyFactory();

  SignumKey.LongKeyFactory<Account.Balance> getAccountBalanceKeyFactory();

  Collection<Account.RewardRecipientAssignment> getAccountsWithRewardRecipient(Long recipientId);

  Collection<Account.AccountAsset> getAssets(int from, int to, Long id);

  Account.AccountAsset getAccountAsset(Long accountId, Long assetId);

  Collection<Account.AccountAsset> getAssetAccounts(Asset asset, boolean ignoreTreasury, long minimumQuantity, boolean unconfirmed, int from, int to);

  // returns true iff:
  // this.publicKey is set to null (in which case this.publicKey also gets set to key)
  // or
  // this.publicKey is already set to an array equal to key
  boolean setOrVerify(Account acc, byte[] key, int height);
}

File: src/main/kotlin/brs/db/store/AccountStore.kt
package brs.db.store

import brs.Account
import brs.db.BurstKey
import brs.db.VersionedBatchEntityTable
import brs.db.VersionedEntityTable

/**
 * Interface for Database operations related to Accounts
 */
interface AccountStore {

    val accountTable: VersionedBatchEntityTable<Account>

    val rewardRecipientAssignmentTable: VersionedEntityTable<Account.RewardRecipientAssignment>

    val rewardRecipientAssignmentKeyFactory: BurstKey.LongKeyFactory<Account.RewardRecipientAssignment>

    val accountAssetKeyFactory: BurstKey.LinkKeyFactory<Account.AccountAsset>

    val accountAssetTable: VersionedEntityTable<Account.AccountAsset>

    val accountKeyFactory: BurstKey.LongKeyFactory<Account>

    fun getAssetAccountsCount(assetId: Long): Int

    fun getAccountsWithRewardRecipient(recipientId: Long?): Collection<Account.RewardRecipientAssignment>

    fun getAssets(from: Int, to: Int, id: Long?): Collection<Account.AccountAsset>

    fun getAssetAccounts(assetId: Long, from: Int, to: Int): Collection<Account.AccountAsset>

    fun getAssetAccounts(assetId: Long, height: Int, from: Int, to: Int): Collection<Account.AccountAsset>
    // returns true iff:
    // this.publicKey is set to null (in which case this.publicKey also gets set to key)
    // or
    // this.publicKey is already set to an array equal to key
    fun setOrVerify(acc: Account, key: ByteArray, height: Int): Boolean
}
